staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller_test.go

```diff
@@ -28,172 +32,597 @@ import (
+					AppOwnerReferences(ownerName).AllNodes(true).
+					Driver(driveName).Devices([]resourceapi.Device{{Name: deviceName}}).
+					Pool(resourceapi.ResourcePool{Name: poolName, Generation: 1, ResourceSliceCount: 1}).Obj(),
+			},
+		},
+		"add-one-network-device-some-nodes": {
+			initialObjects: []runtime.Object{},
+			poolName:       poolName,
+			inputDriverResources: &DriverResources{
+				Pools: map[string]Pool{
+					poolName: {
+						NodeSelector: nodeSelector,
+						Slices:       []Slice{{Devices: []resourceapi.Device{{Name: deviceName}}}},
+					},
+				},
+			},
+			expectedResourceSlices: []resourceapi.ResourceSlice{
+				*MakeResourceSlice().Name(generatedName1).GenerateName(generateName).
+					AppOwnerReferences(ownerName).NodeSelector(nodeSelector).
+					Driver(driveName).Devices([]resourceapi.Device{{Name: deviceName}}).
+					Pool(resourceapi.ResourcePool{Name: poolName, Generation: 1, ResourceSliceCount: 1}).Obj(),
+			},
+		},
+		"update-node-selector": {
+			initialObjects: []runtime.Object{
+				MakeResourceSlice().Name(resourceSlice1).UID(resourceSlice1).
+					AppOwnerReferences(ownerName).NodeSelector(nodeSelector).
+					Driver(driveName).Devices([]resourceapi.Device{{Name: deviceName}}).
+					Pool(resourceapi.ResourcePool{Name: poolName, Generation: 1, ResourceSliceCount: 1}).Obj(),
+			},
+			poolName: poolName,
+			inputDriverResources: &DriverResources{
+				Pools: map[string]Pool{
+					poolName: {
+						NodeSelector: otherNodeSelector,
+						Slices:       []Slice{{Devices: []resourceapi.Device{{Name: deviceName}}}},
+					},
+				},
+			},
+			expectedResourceSlices: []resourceapi.ResourceSlice{
+				*MakeResourceSlice().Name(resourceSlice1).UID(resourceSlice1).ResourceVersion("1").
+					AppOwnerReferences(ownerName).NodeSelector(otherNodeSelector).
+					Driver(driveName).Devices([]resourceapi.Device{{Name: deviceName}}).
+					Pool(resourceapi.ResourcePool{Name: poolName, Generation: 1, ResourceSliceCount: 1}).Obj(),
+			},
 		},
 	}
 
 	for name, test := range testCases {
 		t.Run(name, func(t *testing.T) {
 			_, ctx := ktesting.NewTestContext(t)
-			kubeClient := fake.NewSimpleClientset(test.initialObjects...)
+			inputObjects := make([]runtime.Object, 0, len(test.initialObjects)+len(test.initialOtherObjects))
+			for _, initialOtherObject := range test.initialOtherObjects {
+				inputObjects = append(inputObjects, initialOtherObject.DeepCopyObject())
+			}
+			for _, initialObject := range test.initialObjects {
+				if _, ok := initialObject.(*resourceapi.ResourceSlice); !ok {
+					t.Fatalf("test.initialObjects have to be of type *resourceapi.ResourceSlice")
+				}
+				inputObjects = append(inputObjects, initialObject.DeepCopyObject())
+			}
+			kubeClient := createTestClient(inputObjects...)
 			owner := Owner{
 				APIVersion: "v1",
 				Kind:       "Node",
-				Name:       node,
+				Name:       ownerName,
 				UID:        test.nodeUID,
 			}
+			if test.nodeUID == "" {
```

Comment:
Will our current owner be a resource object other than node?

Reply:
Included in https://github.com/kubernetes/kubernetes/compare/a481793493efdf4267f65c724d7aed71d6fed0e1..0d03fb153e5ef5928e1de18e1ec096c0932c734a

Reply:
Now I need to rebase...